File: developpement/py_lora_mac/lora_phy.py
import serial
import queue
import threading
from enum import Enum, auto, unique
from dataclasses import dataclass
import logging
import time
import csv
import random

log = logging.getLogger("LoRa_ROOT.PHY")

HEADER_SIZE = 16 #NUmber of hexadecimal character in the header

K_FLAG_SHIFT = 7
NEXT_FLAG_SHIFT = 6


ENABLE_STAT = True
TEST_LOSS = False
LOSS_PROBABILITY = 0 #in [0, 1]

@unique
class MacCommand(Enum):
    """The MAC commands available for LoRaMAC."""

    JOIN = 0
    JOIN_RESPONSE = 1
    DATA = 2
    ACK = 3
    QUERY = 4


@unique
class UartCommand(Enum):
    """The UART commands used with the RN2383."""

    MAC_PAUSE = "mac pause"  # pause mac layer
    RX = "radio rx "  # receive mode
    TX = "radio tx "  # transmit data
    SLEEP = "sys sleep "  # system sleep

    """
    The modulation method.
    Values can be: lora, fsk.
    Default: lora
    """
    SET_MOD = "radio set mod "  # set radio mode (fsk or lora)

    """
    The frequency in Hz.
    From 433050000 to 434790000 or from 863000000 to 870000000
    Default: 868100000
    """
    SET_FREQ = "radio set freq "

    """
    The operating bandwidth in KHz.
    Values can be: 125, 250, 500.
    Default: 125
    """
    SET_BW = "radio set bw "

    """
    The transceiver output power.
    From -3 to 15.
    Default: 1
    """
    SET_PWR = "radio set pwr "

    """
    The spreading factor.
    Values can be: sf7, sf8, sf9, sf10, sf11 or sf12.
    Default: sf12
    """
    SET_SF = "radio set sf "

    """
    The coding rate.
    Values can be: 4/5, 4/6, 4/7, 4/8.
    Default: 4/5
    """
    SET_CR = "radio set cr "

    """
    The time-out length for the Watchdog Timer in milliseconds.
    From 0 to 4294967295. 0 disable this functionality.
    Default: 15000
    """
    SET_WDT = "radio set wdt "


@unique
class UartResponse(Enum):
    """The possible UART responses for the RN2483."""

    OK = "ok"
    INVALID_PARAM = "invalid_param"
    RADIO_ERR = "radio_err"
    RADIO_RX = "radio_rx"
    BUSY = "busy"
    RADIO_TX_OK = "radio_tx_ok"
    U_INT = "4294967245" # The response to MAC PAUSE
    NONE = "none"


@dataclass(frozen=True)
class LoraAddr:
    """A LoRaMAC address

    The format of a LoRaMAC adress is the following (size in bits):
                |<---8-->|<---16-->|
                | prefix | node-id |

    Attributes:
        prefix: An integer which is the prefix of the address.
        node_id: An integer chich is the node id of the node.
    """

    prefix: int
    node_id: int

    def toHex(self) -> str:
        """Serialize the adress to a string in hexadecimal.

        Returns:
            str: The serialized address
        """
        return "%02X" % self.prefix + "%04X" % self.node_id

    def __str__(self):
        return str(self.prefix) + ":" + str(self.node_id)


@dataclass
class LoraFrame:
    """A LoRaMAC frame

        The format of a LoRaMAC frame is the following (size in bits):

    |<---24---->|<----24--->|<-1->|<-1-->|<---2--->|<--4--->|<--8--->|<(2040-64=1976)>|
    | dest addr |  src addr |  k  | next | reserved|command |  seq   |     payload    |

        Attributes:
            src_addr: The source address
            dest_addr: The Destination address
            command: The MAC command (c.f. MacCommand)
            payload: The payload. Must be a string in hexadecimal
            seq: The sequence number
            k: True if the frame need an ack, False otherwise
            has_next: True true if another frame follows it, False otherwise. Only used for downward traffic
    """

    src_addr: LoraAddr
    dest_addr: LoraAddr
    command: MacCommand
    payload: str
    seq: int = 0
    k: bool = False
    has_next: bool = False

    def toHex(self) -> str:
        """Serialize the frame.

        Returns:
            str: The serialized frame
        """

        # create flags and MAC command
        f_c = 0
        f_c |= self.k << K_FLAG_SHIFT
        f_c |= self.has_next << NEXT_FLAG_SHIFT
        f_c |= self.command.value

        # check that the size of the payload is even
        if len(self.payload) % 2 != 0:
            # The size must be even because one character is 4 bits and we
            # can't send a half a byte
            # if not even add a zero
            self.payload = "0" + self.payload

        return (
            self.src_addr.toHex()
            + self.dest_addr.toHex()
            + ("%02X" % f_c)
            + ("%02X" % self.seq)
            + (self.payload if self.payload else "")
        )

    @staticmethod
    def build(data: str):
        """Deserialize (or build) a LoRaFrame.

        Returns:
            LoraFrame: The frame built from the data.

        """

        if len(data) < HEADER_SIZE:
            return None

        """ extract src addr: most significant 24 bits"""
        prefix_src = int(data[0:2], 16)
        node_id_src = int(data[2:6], 16)

        """ extract dest addr: next 24 bits """
        prefix_dest = int(data[6:8], 16)
        node_id_dest = int(data[8:12], 16)

        """ extract flags an MAC command: 8 bits """
        f_c = int(data[12:14], 16)

        flag_filter = 0x01  # 1 bit
        cmd_filter = 0x0F  # 4 bits

        k = bool((f_c >> K_FLAG_SHIFT) & flag_filter)
        has_next = bool((f_c >> NEXT_FLAG_SHIFT) & flag_filter)

        cmd = MacCommand(f_c & cmd_filter)

        """ extract sequence number: 8 bits """
        seq = int(data[14:16], 16)

        """ extract payload """
        payload = data[16:]

        """ create LoraFrame with computed values"""
        return LoraFrame(
            LoraAddr(prefix_src, node_id_src),
            LoraAddr(prefix_dest, node_id_dest),
            cmd,
            payload,
            seq,
            k,
            has_next,
        )


@dataclass
class UartFrame:
    """An UART paquet

    Attributes:
        expected_response: The expected UART response to this paquet
        data: The sent data
        cmd: The UART command (c.f. UartCommand)
    """

    expected_response: list
    data: str
    cmd: UartCommand
    stat_id: int = -1


class PhyMeter:

    instance = None

    def __init__(self, dest_file):
        self.data_list=[]#[[data, put_time, send_time, diff]]
        self.counter = -1
        self.dest_file = dest_file

    def put(self,data):
        data.stat_id = self._get_id()
        self.data_list.append([data, time.perf_counter_ns(), -1, -1])

    def send(self,data):
        r = self.data_list[data.stat_id]
        r[2]=time.perf_counter_ns()
        r[3]=r[2]-r[1]

    @staticmethod
    def getMeter(dest_file='stat.txt'):
        if PhyMeter.instance == None:
            PhyMeter.instance = PhyMeter(dest_file)
        return PhyMeter.instance

    def _get_id(self):
        self.counter+=1
        return self.counter

    def export_data(self):
        #with open(self.dest_file, "w", newline='') as f:
        writer = csv.writer(open(self.dest_file, "w", newline=''), 'unix')
        writer.writerow(['put time', 'send time', 'delta', 'data'])
        writer.writerow([self.data_list[0][0], self.data_list[-1][1], len(self.data_list), 'general info'])
        for data in self.data_list:
            writer.writerow([data[1], data[2], data[3], str(data[0])])



class LoraPhy:
    """The LoRaMAC PHY layer.

    The PHY layer is the driver for the RN2483.

    """

    def __init__(self, **params):
        self._con = None  # The serial conenction
        #self._port = port  # The serial port
        #self._baudrate = baudrate  # The serial baudrate
        self._params = params
        self._buffer = queue.Queue(self._params.get('txBufSize', 10))  # The TX buffer
        self._rx_buffer = queue.Queue(self._params.get('rxBufSize', 10))  # the RX buffer
        self._can_send = True  # True if the tx process can send, False otherwise
        self._can_send_cond = threading.Condition()  # condition used by the tx_process
        self._last_sended = None  # the last sended frame
        self._tx_lock = threading.Lock()  # lock used for phy_tx()
        self.listen_lock = threading.Lock()
        self._is_listen = False

    def init(self):
        """Init the PHY layer.

        - Set the serial connection
        - Prepare the RN2483 for communications (mac pause, radio set)
        """
        # set serial connection, call send_phy for mac pause et radio set freq
        log.info("Init PHY")
        try:
            self._con = serial.Serial(port=self._params.get('port', "/dev/ttyUSB0"), baudrate=self._params.get('baudrate', 57600))
        except serial.serialutil.SerialException as e:
            log.error(str(e))
            exit()
        tx_thread = threading.Thread(target=self._uart_tx)
        rx_thread = threading.Thread(target=self._uart_rx)

        log.info("Radio configuration: %s", str(self._params))

        # Radio configuration
        self._send_phy(UartFrame([UartResponse.U_INT], "", UartCommand.MAC_PAUSE))
        self._send_phy(UartFrame([UartResponse.OK], self._params.get('mode',"lora"), UartCommand.SET_MOD))
        self._send_phy(UartFrame([UartResponse.OK], self._params.get('frequence',"868100000"), UartCommand.SET_FREQ))
        self._send_phy(UartFrame([UartResponse.OK], self._params.get('bandwidth',"125"), UartCommand.SET_BW))
        self._send_phy(UartFrame([UartResponse.OK], self._params.get('cr',"4/5"), UartCommand.SET_CR))
        self._send_phy(UartFrame([UartResponse.OK], self._params.get('pwr',"1"), UartCommand.SET_PWR))
        self._send_phy(UartFrame([UartResponse.OK], self._params.get('sf',"sf2"), UartCommand.SET_SF))
        
        rx_thread.start()
        tx_thread.start()

        with self._can_send_cond:
            self._can_send_cond.notify_all()

    def phy_send(self, loraFrame: LoraFrame):
        """Method to use to send LoraFrame.

        Prepare the UART paquet from the LoRa frame.

        Args:
            loraFrame (LoraFrame): The frame to sent.
        """

        log.info("MAC send:%s", str(loraFrame))
        self._tx_lock.acquire()
        if TEST_LOSS:
            if not random.random() < LOSS_PROBABILITY:
                log.warning("LOSS TEST: don't send %s", str(loraFrame))
                return
        if loraFrame is None:
            return
        f = UartFrame(
            [UartResponse.RADIO_TX_OK, UartResponse.RADIO_ERR],
            loraFrame.toHex(),
            UartCommand.TX,
        )
        self._send_phy(f)
        self._tx_lock.release()

    def phy_timeout(self, timeout: int):
        """Set the RN2483 radio watchdog timer.

        Args:
            timeout (int): The timeout (in milliseconds)
        """

        log.info("MAC set watchdog timer to %d ms", timeout)
        f = UartFrame([UartResponse.OK], str(timeout), UartCommand.SET_WDT)
        self._send_phy(f)

    def phy_rx(self):
        """Set the radio the reception mode"""

        self.listen_lock.acquire()
        self._is_listen = True
        self.listen_lock.release()
        log.info("MAC switch to reception mode")
        f = UartFrame(
            [UartResponse.RADIO_ERR, UartResponse.RADIO_RX], "0", UartCommand.RX
        )
        self._send_phy(f)

    def listen(self) -> bool:
        result = None
        self.listen_lock.acquire()
        result = self._is_listen
        self.listen_lock.release()
        return result

    def getFrame(self) -> LoraFrame:
        """Get the next received frame.

        This method block until a frame is available.

        Returns:
            LoraFrame: A received frame.
        """
        frame = self._rx_buffer.get()
        return frame

    def _send_phy(self, data: UartFrame) -> bool:
        """Append data to the TX buffer.

        Args:
            data (UartFrame): The UART frame to put in the tx buffer.

        Raises:
            TypeError: If the data type is not UartFrame

        Returns:
            bool: True if the data has been added to the buffer, False otherwise
        """

        if type(data) != UartFrame:
            raise TypeError("Data must be UartFrame. actual type: ", type(data))

        try:
            log.debug("append %s to tx_buf", str(data))
            self._buffer.put(data, block=False)
            if ENABLE_STAT:
                PhyMeter.getMeter().put(data)
        except queue.Full:
            log.warning("TX buffer full")
            return False

        return True

    def _process_response(self, decode_data: str) -> bool:
        """Process an UART response.

        Args:
            data (str): The UART response.

        Returns:
            bool: True if the answer is the one expected, False otherwise.
        """
        #decode_data = data
        log.debug("process uart response: " + decode_data)
        for resp in self._last_sended.expected_response:
            if resp is None:
                continue
            if resp.value in decode_data:  # the response is the one expected
                if resp == UartResponse.RADIO_RX:  # the response is DATA
                    log.info("PHY RX:" + decode_data[10:].strip())
                    #if ENABLE_STAT:
                    #    log.info("STAT: PHY RX: %d", time.monotonic_ns())
                    try:
                        self._rx_buffer.put(
                            LoraFrame.build(decode_data[10:].strip()), block=False
                        )
                    except queue.Full:
                        log.warning("RX buffer full")

                return True
        log.debug("unexpected response")
        return False

    def _uart_rx(self):
        """Method used as Thread to read data from the serial connection."""
        while True:
            data = self._con.readline().strip().decode()
            if ("radio_rx" in data) or ("radio_err" in data):
                self.listen_lock.acquire()
                self._is_listen = False
                self.listen_lock.release()

            log.debug("UART response: %s", data)
            if self._process_response(data):
                # It is the expected response
                # Notify threads waiting for the response
                with self._can_send_cond:
                    self._can_send = True
                    self._can_send_cond.notify_all()

    def _uart_tx(self):
        """Method used as Thread to send data to the serial connection."""
        while True:
            while self._con is None or not self._can_send:
                with self._can_send_cond:
                    self._can_send_cond.wait()
            self._last_sended = self._buffer.get(block=True)
            log.info("PHY TX:" + self._last_sended.cmd.value + self._last_sended.data)
            if ENABLE_STAT:
                PhyMeter.getMeter().send(self._last_sended)
                #if self._last_sended.cmd == UartCommand.TX:
                #    log.info("STAT: PHY TX: %d", time.monotonic_ns())
            self._con.write(
                (self._last_sended.cmd.value + self._last_sended.data + "\r\n").encode()
            )
            self._can_send = False

File: developpement/lorarpl/loraphy.c
#include "contiki.h"

#include "dev/uart.h"
#include "sys/log.h"
#include "sys/mutex.h"

#include <stdbool.h>
#include <stdlib.h>

#include "loraphy.h"
/*---------------------------------------------------------------------------*/
//log configuration
#define LOG_MODULE "LoRa PHY"
#define LOG_LEVEL LOG_LEVEL_INFO
#define LOG_CONF_WITH_COLOR 3

//TX buffer initialization
#define TX_BUF_SIZE 10
static uart_frame_t tx_buffer [TX_BUF_SIZE];
static uint8_t w_i = 0;// index to write in the buffer
static uint8_t r_i = 0;// index to read in the buffer 
static uint8_t tx_buf_size = 0;// current size of the buffer

//Function to send data to. Data from ratio_rx
static int (* handler)( lora_frame_t frame) = NULL;

//Array that contains the expected responses for a sent UART command
//The expected responses, when received, allow the next UART command to be sent
static uart_response_t expected_response[UART_EXP_RESP_SIZE];
const char* uart_response[8]={"ok", "invalid_param", "radio_err", "radio_rx", "busy", "radio_tx_ok", "4294967245", "none"};

//commands that can be sent
//see uart_command_t
const char* uart_command[7]={"mac pause", "radio set mod ", "radio set freq ", "radio set wdt ", "radio rx ", "radio tx ", "sys sleep "};


mutex_t response_mutex;//mutex for expected_response
mutex_t tx_buf_mutex;// mutex for tx_buffer

static process_event_t new_tx_frame_event;//event that signals to the TX process that a new frame is available to be sent //V
static process_event_t can_send_event;//event that signals to the TX process that the correct UART response was received//V

static bool can_send = true;

PROCESS(ph_rx, "LoRa-PHY rx process");//V
PROCESS(ph_tx, "LoRa-PHY tx process");//V


/*---------------------------------------------------------------------------*/

void print_uart_frame(uart_frame_t *frame){
    printf("{ cmd:%s ", uart_command[frame->cmd]);
    printf("resp: [");
    uart_response_t *f_expected_response = frame->expected_response;
    for(int i=0;i<UART_EXP_RESP_SIZE-1;i++){
        printf("%s, ",uart_response[f_expected_response[i]]);
    }
    printf("%s] }",uart_response[f_expected_response[UART_EXP_RESP_SIZE-1]]);
    
}
/*build lora_frame_t from char**/
int parse(lora_frame_t *dest, char *data){

    if(strlen(data) < HEADER_SIZE){
        return 1;
    }
    
    lora_frame_t result;
    
    char prefix_c[2];
    char id_c[4];
    
    uint8_t prefix;
    uint16_t id;
    lora_addr_t src_addr;
    lora_addr_t dest_addr;

    /*extract src addr*/
    memcpy(prefix_c, data, 2);
    data = data+2;
    memcpy(id_c, data, 4);
    data = data+4;

    prefix = (uint8_t) strtol(prefix_c, NULL, 16);
    id = (uint16_t) strtol(id_c, NULL, 16);

    src_addr.prefix = prefix;
    src_addr.id=id; 
    result.src_addr = src_addr;

    /*extract dest addr*/
    memcpy(prefix_c, data, 2);
    data = data+2;
    memcpy(id_c, data, 4);
    data = data+4;

    prefix = (uint8_t) strtol(prefix_c, NULL, 16);
    id = (uint16_t) strtol(id_c, NULL, 16);

    dest_addr.prefix = prefix;
    dest_addr.id = id;
    result.dest_addr = dest_addr;

    /*extact flags and command*/
    char cmd[2];
    memcpy(cmd, data, 2);
    data = data+2;
    uint8_t i_cmd = (uint8_t)strtol(cmd, NULL, 16);

    uint8_t flag_filter = 0x01;
    uint8_t command_filter = 0x0F;

    bool k    = (bool)((i_cmd >> 7) & flag_filter);
    bool seq  = (bool)((i_cmd >> 6) & flag_filter);
    bool next = (bool)((i_cmd >> 5) & flag_filter);
    
    mac_command_t command = (uint8_t)( i_cmd & command_filter );
    
    result.k = k;
    result.command = command;
    result.seq = seq;
    result.next = next;

    /*extract payload*/
    result.payload = data;
    memcpy(dest, &result, sizeof(lora_frame_t));
    LOG_DBG("created frame a LoRa frame\n");
    
    return 0;
}

/*convert lora_frame_t to hex*/
int to_frame(lora_frame_t *frame, char *dest){
    
    char result[HEADER_SIZE+PAYLOAD_MAX_SIZE]="";

    /*create src and dest addr*/
    char src_addr[6];
    char dest_addr[6];
    
    sprintf(src_addr, "%02X%04X", frame->src_addr.prefix, frame->src_addr.id);
    sprintf(dest_addr, "%02X%04X", frame->dest_addr.prefix, frame->dest_addr.id);
    
    /*create flags and MAC command*/
    char flags_command[2];

    uint8_t f_c = 0;
    if(frame->k){
        f_c = f_c | K_FLAG;
    }
    if(frame->seq){
        f_c = f_c | SEQ_FLAG;
    }
    if(frame->next){
        f_c = f_c | NEXT_FLAG;
    }
    f_c = f_c | ((uint8_t) frame->command);

    sprintf(flags_command, "%02X", f_c);   
    
    /* concat all computed values to result */
    strcat(result, src_addr);
    strcat(result, dest_addr);
    strcat(result, flags_command);
    
    /* create payload */
    int real_frame_size = HEADER_SIZE;
    if(frame->payload != NULL){
        real_frame_size = real_frame_size + strlen(frame->payload);
    } 
    //int size = HEADER_SIZE+payload_size+9;//todo + 9 why ?
    if(real_frame_size>HEADER_SIZE){
        strcat(result, frame->payload);
    }
    
    /*copy result to dest */
    memcpy(dest, &result, real_frame_size+1);
    LOG_DBG("TO HEX frame: %s\n",result);
    return 0;
}

void write_uart(char *s){
    LOG_INFO("Write UART:%s\n", s);
    while(*s != 0){
        uart_write_byte(UART, *s++);
    }
    uart_write_byte(UART, '\r');
    uart_write_byte(UART, '\n');
}

void phy_register_listener(int (* listener)(lora_frame_t frame)){
    handler = listener;
}


/*---------------------------------------------------------------------------*/

void process_command(unsigned char *command){
    LOG_INFO("UART response:%s\n", command);

    lora_frame_t frame;

    while(!mutex_try_lock(&response_mutex)){}
    for(int i=0;i<UART_EXP_RESP_SIZE;i++){
        LOG_DBG("compare response with:%s\n",uart_response[expected_response[i]]);
        if(strstr((const char*)command, uart_response[expected_response[i]]) != NULL){
            /*the UART response is the exoetced response*/
            if(expected_response[i] == RADIO_RX && parse(&frame, (char*)(command+10))==0){
                /*receive data -> transmit to MAC layer*/
                handler(frame);
            }
            /* signal to the tx process that the next frame can be sent*/
            LOG_DBG(" expected !\n");
            process_post(&ph_tx, can_send_event, NULL);
            break;
        }
    }
    mutex_unlock(&response_mutex);

}

/**
 * Callback function that receive bytes from UART
 **/
int uart_rx(unsigned char c){
    static unsigned char buf [FRAME_SIZE];
    static unsigned short index = 0;
    static bool cr = false;

    if(c == '\r'){
      cr = true;
    }else if(c == '\n'){
      if(cr==true){
        process_command(buf);
        index = 0;
        cr = false;
        memset(buf, 0, FRAME_SIZE*sizeof(char));
      }
    }
    if((int)c != 254 && (int)c != 248 && c!='\n' && (int)c != 192 && (int)c != 240 && c!='\r'){
        buf[index++] = c;
    }
    return 0;
}

int uart_tx(uart_frame_t uart_frame){
    LOG_DBG("enter UART_TX\n");
    while(!mutex_try_lock(&tx_buf_mutex)){}
    if(tx_buf_size < TX_BUF_SIZE){
        tx_buffer[w_i] = uart_frame;
        tx_buf_size ++;
        w_i = (w_i+1)%TX_BUF_SIZE;
        LOG_DBG("append ");
        LOG_DBG_UFRAME(&uart_frame);
        LOG_DBG(" to UART TX buffer\n");
    }
    mutex_unlock(&tx_buf_mutex);
    process_post(&ph_tx, new_tx_frame_event, NULL);
    return 0;//TODO
}

void phy_init(){
    //add config command to tx_buf
    //send event to process
    LOG_INFO("Init LoRa PHY\n");//V

    //create events
    new_tx_frame_event = process_alloc_event();//V
    can_send_event = process_alloc_event();//V
    
    //start process
    process_start(&ph_rx, NULL);//V
    process_start(&ph_tx, NULL);//V

    //send initialisation UART commands
    uart_frame_t mac_pause = {MAC_PAUSE, STR, {.s=""}, {U_INT, NONE}};
    uart_frame_t set_freq = {SET_FREQ, STR, {.s="868100000"}, {OK, NONE}};

    uart_tx(mac_pause);
    uart_tx(set_freq);

}

int phy_tx(lora_frame_t frame){
    uart_frame_t uart_frame = {
        TX,
        LORA,
        {.lora_frame = frame},
        {RADIO_TX_OK,RADIO_ERR}
    };
    uart_tx(uart_frame);
    return 0;

}

int phy_timeout(int timeout){
    if(timeout < 0 || timeout > 4294967295 ){
        return 1;
    }

    uart_frame_t uart_frame = {
        SET_WDT,
        INT,
        {.d = timeout},
        {OK, NONE}
    };
    uart_tx(uart_frame);
    return 0;
}

int phy_sleep(int duration){
    if(duration < 100 || duration > 4294967296 ){
        return 1;
    }

    uart_frame_t uart_frame = {
        SLEEP,
        INT,
        {.d = duration},
        {OK, NONE}
    };
    uart_tx(uart_frame);
    return 0;
}

int phy_rx(){
    uart_frame_t uart_frame = {
        RX,
        STR,
        {.s = "0"},
        {RADIO_ERR, RADIO_RX}
    };
    uart_tx(uart_frame);
    return 0;
}

/*---------------------------------------------------------------------------*/
// process
PROCESS_THREAD(ph_rx, ev, data){//V
    PROCESS_BEGIN();
    
    //UART configuration
    uart_init(UART);
    uart_set_input(UART, &uart_rx);
    
    PROCESS_END();//V
}

bool buf_empty(){
    bool result;
    while(!mutex_try_lock(&tx_buf_mutex)){}
    result = tx_buf_size == 0;
    mutex_unlock(&tx_buf_mutex);
    return result;
}

PROCESS_THREAD(ph_tx, ev, data){//V
    
    uart_frame_t uart_frame;
    bool buf_empty_var;
    
    PROCESS_BEGIN();//V

    //main process
    while (true){
        buf_empty_var = buf_empty();
        if(buf_empty_var){
            PROCESS_WAIT_EVENT_UNTIL(ev==new_tx_frame_event);
        }
        
        if(!can_send){//to be sure. but should never happen.
            continue;
        }
        
        //acquire mutex for buffer
        while(!mutex_try_lock(&tx_buf_mutex)){}
        
        //get next uart frame to send
        uart_frame = tx_buffer[r_i];
        tx_buf_size --;
        r_i = (r_i+1)%TX_BUF_SIZE;
        mutex_unlock(&tx_buf_mutex);
        
        //acquire mutex for expected_response
        while(!mutex_try_lock(&response_mutex)){}
        
        //update expected response
        for(int i=0;i<UART_EXP_RESP_SIZE;i++){
            expected_response[i] = uart_frame.expected_response[i];
        }
        mutex_unlock(&response_mutex);

        can_send = false;
        
        char result[FRAME_SIZE]="";
        if(uart_frame.type == STR){
            sprintf(result, "%s%s", uart_command[uart_frame.cmd], uart_frame.data.s);
        }else if(uart_frame.type == LORA){
            strcat(result, uart_command[uart_frame.cmd]);
            to_frame(&(uart_frame.data.lora_frame), result+strlen(result));
        }else if(uart_frame.type == INT){
            sprintf(result, "%s%d", uart_command[uart_frame.cmd], uart_frame.data.d);
        }

        write_uart(result);
        while(!can_send){
            PROCESS_WAIT_EVENT();
            if(ev == can_send_event){
                can_send = true;
                
            }
        }
    }

    PROCESS_END();//V
}